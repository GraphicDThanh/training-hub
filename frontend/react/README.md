# React Redux Training
[Plan doc](https://docs.google.com/document/d/1lYywN0RIhrwsyFZ8uPZuaN5QM78mYuXFD608vS8YEbM/edit?pli=1#)
## ES6
## React Basic
### Practice 1 - examples React
## React Advanced
### Practice 2
## Redux Basic
## Redux Advanced
## Redux Recipes
