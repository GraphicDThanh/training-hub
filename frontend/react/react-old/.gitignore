node_modules
out
dist
.DS_Store
.git
.docpad.db
vendor/
.vagrant
bower-components/
bower_components/
npm-debug.log
.gitignore~
vagrant/Vagrantfile~
build/
