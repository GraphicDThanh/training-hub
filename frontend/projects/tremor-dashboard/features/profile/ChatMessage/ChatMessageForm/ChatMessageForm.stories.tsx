import { Meta, StoryObj } from "@storybook/react";

// Components
import ChatMessageForm from "./ChatMessageForm";

const mockHandleSendMessage = async (message: string) => {};

const meta = {
  title: "Components/Profiles/ChatMessageForm",
  tags: ["autodocs"],
  component: ChatMessageForm,
  argTypes: {
    onSendMessage: { description: "The function handler send message" },
  },
} as Meta<typeof ChatMessageForm>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: { onSendMessage: mockHandleSendMessage },
};
