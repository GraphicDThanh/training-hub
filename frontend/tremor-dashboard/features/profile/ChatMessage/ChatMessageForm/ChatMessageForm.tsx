"use client";

import { memo, useCallback } from "react";
import { Flex } from "@tremor/react";
import { Controller, useForm } from "react-hook-form";

// Constants
import { VARIANT_BUTTON } from "@/constants";

// Types
import { Message } from "@/types";

// Icons
import { IoIosSend } from "react-icons/io";

// Components
import { Button, InputField } from "@/components";

const ChatMessageForm = ({
  onSendMessage,
}: {
  onSendMessage: (message: string) => Promise<void>;
}) => {
  const { control, handleSubmit, reset } = useForm<Message>({
    defaultValues: { message: "" },
  });

  const handleSendMessage = useCallback(
    async (value: Message) => {
      const message = value.message.trim();

      if (!message) return;

      reset({ message: "" });

      await onSendMessage(message);
    },
    [reset, onSendMessage],
  );

  return (
    <form id="sendMessageForm" onSubmit={handleSubmit(handleSendMessage)}>
      <Flex className="items-center gap-3 p-3 bg-greyish dark:bg-primary">
        <Controller
          name="message"
          control={control}
          render={({ field: { value, onChange } }) => (
            <InputField
              id="message"
              placeholder="Type message"
              value={value}
              onChange={onChange}
              data-testid="chat-message-input"
            />
          )}
        />
        <Button
          type="submit"
          variant={VARIANT_BUTTON.SECONDARY}
          data-testid="conversation-sent-btn"
          additionalClass="bg-transparent border-0 p-0 dark:bg-transparent dark:hover:bg-transparent">
          <IoIosSend className="text-xl text-primary dark:text-white " />
        </Button>
      </Flex>
    </form>
  );
};

export default memo(ChatMessageForm);
