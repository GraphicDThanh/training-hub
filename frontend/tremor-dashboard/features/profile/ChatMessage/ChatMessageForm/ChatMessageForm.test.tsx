import { fireEvent, render, waitFor } from "@testing-library/react";

import "@testing-library/jest-dom";
import ChatMessageForm from "./ChatMessageForm";

const message = "Hi!";

describe("ChatMessageForm", () => {
  const mockHandleSendMessage = jest.fn();

  const renderComponent = () =>
    render(<ChatMessageForm onSendMessage={mockHandleSendMessage} />);

  it("Should match snapshot", () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  it("Should not trigger onSendMessage when user yet type but submit", async () => {
    const { getByRole } = renderComponent();

    const button = getByRole("button");
    fireEvent.click(button);
    await waitFor(() => expect(mockHandleSendMessage).not.toHaveBeenCalled());
  });

  it("Should trigger onSendMessage when user type and submit", async () => {
    const { getByTestId, getByRole } = renderComponent();

    const input = getByTestId("chat-message-input");
    fireEvent.change(input, { target: { value: message } });
    expect(input).toHaveValue(message);

    const button = getByRole("button");
    fireEvent.click(button);
    await waitFor(() =>
      expect(mockHandleSendMessage).toHaveBeenNthCalledWith(1, message),
    );
  });
});
