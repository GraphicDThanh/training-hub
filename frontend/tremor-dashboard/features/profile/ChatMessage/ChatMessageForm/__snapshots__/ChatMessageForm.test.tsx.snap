// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatMessageForm Should match snapshot 1`] = `
<div>
  <form
    id="sendMessageForm"
  >
    <div
      class="tremor-Flex-root flex w-full flex-row justify-between items-center gap-3 p-3 bg-greyish dark:bg-primary"
    >
      <div
        class="w-full"
      >
        <label
          class="text-grey text-sm dark:text-lighter"
          for="message"
        />
        <div
          class="tremor-Flex-root flex w-full flex-row justify-between items-center relative"
        >
          <div
            class="w-full"
          >
            <input
              class="block py-1.5 px-0 w-full text-sm text-primary bg-transparent border-0 border-b-2 border-gray-300 appearance-none dark:text-white dark:border-gray-600 dark:focus:border-gray-500 focus:outline-none focus:ring-0 focus:border-gray-600 peer disabled:cursor-not-allowed disabled:opacity-50 "
              data-testid="chat-message-input"
              id="message"
              placeholder="Type message"
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
      <button
        class="tremor-Button-root shrink-0 inline-flex justify-center items-center group font-medium outline-none rounded-tremor-default shadow-tremor-input dark:shadow-dark-tremor-input text-sm dark:text-dark-tremor-brand-inverted hover:border-tremor-brand-emphasis dark:hover:border-dark-tremor-brand-emphasis !rounded-full border-secondary dark:border-secondary text-secondary hover:text-secondary hover:opacity-75 hover:bg-transparent box-shadow-transparent bg-transparent border-0 p-0 dark:bg-transparent dark:hover:bg-transparent"
        data-testid="conversation-sent-btn"
        type="submit"
      >
        <span
          class="tremor-Button-text text-tremor-default whitespace-nowrap"
        >
          <svg
            class="text-xl text-primary dark:text-white "
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 512 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M435.9 64.9l-367.1 160c-6.5 3.1-6.3 12.4.3 15.3l99.3 56.1c5.9 3.3 13.2 2.6 18.3-1.8l195.8-168.8c1.3-1.1 4.4-3.2 5.6-2 1.3 1.3-.7 4.3-1.8 5.6L216.9 320.1c-4.7 5.3-5.4 13.1-1.6 19.1l64.9 104.1c3.2 6.3 12.3 6.2 15.2-.2L447.2 76c3.3-7.2-4.2-14.5-11.3-11.1z"
            />
          </svg>
        </span>
      </button>
    </div>
  </form>
</div>
`;
